CONTRIBUTING.md

```diff
@@ -2,11 +2,75 @@
-
-The verification process is as follow:
-- Create a branch with the changes in the GitHub actions. Push this branch to the GitHub repo.
-- Create branches out of current master in the projects mentioned in [README.md](README.md). In those branches, point the GitHub actions to your new branch.
-- Create pull requests, marked as `do-not-merge` in the projects so that CI is triggered with your new changes.
-- Once working, submit a PR to the GitHub actions repo.
+The projects mentioned in [README.md](README.md) are considered critical users of this repo.
+When introducing a change in this repo, care should be taken to check that those changes will not break the CI process of the critical users.
+
+To submit a change please follow the guidelines described below.
+
+## Basics
+### Pointing to a specific action
+The syntax for using a public action in a subdirectory is as follows ([reference](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-a-public-action-in-a-subdirectory)):
+```
+uses: {owner}/{repo}/{path}@{ref}
+```
+
+For example:
+```
+uses: mfem/github-actions/build-hypre@v2.4
+```
+`{ref}` can be *any* of the following:
+- branch
+- tag
+- commit hash
+
+## Guidelines
+To submit a change please follow these guidelines.
+Because this involves making changes in multiple repos, each step will be marked with the repo it is referencing:
+ - `actions` refers to making a change in this repo: `mfem/github-actions`
+ - `project` refers to making a change in the repo with a workflow invoking an action, e.g. one of the projects mentioned in [README.md](README.md)
+
+### Steps
+1. `actions` Create a branch from `master` with a descriptive name, e.g. `add-single-precision-ci`
+2. `actions` Make updates to actions and push changes
+3. `project` Create a new branch and in that branch, change `{ref}` in the workflows you'd like to test so that it points to your feature branch, e.g.
+> `mfem/.github/workflows/build-and-tests.yml`
+> ```diff
+> - uses: mfem/github-actions/build-mfem@v2.4
+> + uses: mfem/github-actions/build-mfem@add-single-precision-ci
+> ```
+4. `project` To test the updated workflow, create a pull request and label it with `DO-NOT-MERGE`; this will trigger the CI to run with your new changes
+
+> **Note**
+>
+> Adding an option to use manually trigger can be useful for debugging. This can be done by adding `workflow_dispatch` to the list of triggers
+> ```
+> on:
+>   workflow_dispatch
+> ```
+> The option to manually trigger will not appear until the workflow has previously been triggered via something like a PR or push.
+
+5. `actions` Continue to develop and test your action; because the `{ref}` in your `project` workflow is pointing to your branch, re-running the `project` CI will always point to the latest commit.
+
+6. `actions` Once your changes are finalized, submit a PR to pull your branch into `master`
```

Comment:
Would be nice to add a note that the changelog should be updated, because there has not been any culture of that. The last entry is version 1.0 😂 .

Reply:
Great suggestions. I pushed another commit adding both of them in, let me know what you think.

I don't think it'd be that useful to comb through old commits and try to infer what old changes are, but hopefully we can update the changelog at least for `v2.6` onwards and maybe `v2.5`

Reply:
Thank you for all the feedback 🙏😊

Reply:
Thanks for doing this! Do you want to update that changelog here? It would make sense to me as we are changing the rules here...

Reply:
Ok, done 😉 (Feel free to revert it if you do not like it here)